server/src/main/java/org/opensearch/index/codec/composite/Composite99DocValuesReader.java

```diff
@@ -29,11 +60,123 @@
+                Composite99DocValuesFormat.VERSION_CURRENT,
+                readState.segmentInfo.getId(),
+                readState.segmentSuffix
+            );
+
+            // initialize data input
+            metaIn = readState.directory.openChecksumInput(metaFileName, readState.context);
+            Throwable priorE = null;
+            try {
+                CodecUtil.checkIndexHeader(
+                    metaIn,
+                    Composite99DocValuesFormat.META_CODEC_NAME,
+                    Composite99DocValuesFormat.VERSION_START,
+                    Composite99DocValuesFormat.VERSION_CURRENT,
+                    readState.segmentInfo.getId(),
+                    readState.segmentSuffix
+                );
+
+                while (true) {
+
+                    // validate magic marker
+                    long magicMarker = metaIn.readLong();
+                    if (magicMarker == -1) {
+                        logger.info("EOF reached for composite index metadata");
+                        break;
+                    } else if (magicMarker < 0) {
+                        throw new CorruptIndexException("Unknown token encountered: " + magicMarker, metaIn);
+                    } else if (MAGIC_MARKER != magicMarker) {
+                        logger.error("Invalid composite field magic marker");
+                        throw new IOException("Invalid composite field magic marker");
+                    }
+
+                    // construct composite index metadata
+                    CompositeIndexMetadata compositeIndexMetadata = new CompositeIndexMetadata(metaIn);
+                    String compositeFieldName = compositeIndexMetadata.getCompositeFieldName();
+                    compositeFieldInfos.add(
+                        new CompositeIndexFieldInfo(compositeFieldName, compositeIndexMetadata.getCompositeFieldType())
+                    );
+
+                    switch (compositeIndexMetadata.getCompositeFieldType()) {
+                        case STAR_TREE:
+                            StarTreeMetadata starTreeMetadata = compositeIndexMetadata.getStarTreeMetadata();
+                            Tree starTree = new StarTree(dataIn, starTreeMetadata);
+                            starTreeMap.put(compositeFieldName, starTree);
+                            compositeIndexMetadataMap.put(compositeFieldName, compositeIndexMetadata);
+
+                            List<Integer> dimensionFieldNumbers = starTreeMetadata.getDimensionFieldNumbers();
+                            List<FieldInfo> dimensions = new ArrayList<>();
+                            for (Integer fieldNumber : dimensionFieldNumbers) {
+                                dimensions.add(readState.fieldInfos.fieldInfo(fieldNumber));
+                            }
+
+                            // initialize star-tree doc values producer
+                            StarTree99DocValuesProducer starTreeDocValuesProducer = new StarTree99DocValuesProducer(
```

Comment:
Major : this must be shared across all composite fields ( since we are using one consumer for all fields during write )